File: back-end/app/core/cpf.py
from __future__ import annotations

import re


def normalize_cpf(value: str) -> str:
    return re.sub(r"\D+", "", value or "")


def is_valid_cpf(value: str) -> bool:
    cpf = normalize_cpf(value)
    if len(cpf) != 11:
        return False
    if cpf == cpf[0] * 11:
        return False

    first_digit = _calculate_check_digit(cpf[:9], start_factor=10)
    second_digit = _calculate_check_digit(cpf[:10], start_factor=11)
    return cpf[-2:] == f"{first_digit}{second_digit}"


def _calculate_check_digit(base: str, *, start_factor: int) -> int:
    total = sum(int(char) * factor for char, factor in zip(base, range(start_factor, 1, -1), strict=True))
    remainder = total % 11
    return 0 if remainder < 2 else 11 - remainder

File: Front-end/src/app/lib/cpf.ts
export function normalizeCpf(value: string) {
  return (value || "").replace(/\D+/g, "").slice(0, 11);
}

export function formatCpf(value: string) {
  const digits = normalizeCpf(value);
  const parts = [
    digits.slice(0, 3),
    digits.slice(3, 6),
    digits.slice(6, 9),
    digits.slice(9, 11),
  ].filter((part) => part.length > 0);

  if (parts.length <= 1) {
    return parts[0] ?? "";
  }
  if (parts.length === 2) {
    return `${parts[0]}.${parts[1]}`;
  }
  if (parts.length === 3) {
    return `${parts[0]}.${parts[1]}.${parts[2]}`;
  }
  return `${parts[0]}.${parts[1]}.${parts[2]}-${parts[3]}`;
}

export function isValidCpf(value: string) {
  const cpf = normalizeCpf(value);
  if (cpf.length !== 11) {
    return false;
  }
  if (/^(\d)\1{10}$/.test(cpf)) {
    return false;
  }

  const firstDigit = calculateDigit(cpf.slice(0, 9), 10);
  const secondDigit = calculateDigit(cpf.slice(0, 10), 11);
  return cpf.endsWith(`${firstDigit}${secondDigit}`);
}

function calculateDigit(base: string, startFactor: number) {
  const sum = base
    .split("")
    .reduce((total, char, index) => total + Number(char) * (startFactor - index), 0);
  const remainder = sum % 11;
  return remainder < 2 ? 0 : 11 - remainder;
}
